rfcs/2024-07-21-launcher.md

```diff
@@ -0,0 +1,179 @@
+  refreshInterval: 60 * 60, // in seconds, 1 hour
+});
+```
+
+It currently does not have any option for request timeouts and retry logic.
+
+### Challenges
+
+While Featurevisor SDKs allow refreshing datafile content without having to restart the application, options like `refreshInterval` are still hardcoded by application developers.
+
+To make any changes there, we still have to rely on making application code changes and deploy them again. For native apps like iOS and Android, this can be more challenging since we have to wait longer for this kind of change to take impact for all users.
+
+## Detailed design
+
+There are two types of configuration to be fetched by clients here:
+
+- **Datafile**: the generated JSON file which packs all the feature configurations in
+- **Launcher config** (new): a new kind of JSON file containing configuration for setting up the SDK in apps
+
+### Example usage
+
+Setting up the SDK using launcher config URL only:
+
+```js
+import { createInstance } from "@featurevisor/sdk";
+
+const f = createInstance({
+  launcherUrl: "https://cdn.yoursite.com/production/launcher-tag-web.json",
+});
+```
+
+Contents of the fetched launcher config JSON file:
+
+```js
+// GET https://cdn.yoursite.com/production/launcher-tag-web.json
+
+{
+  // Datafile fetching specific configs
+  datafile: {
+    // required
+    url: "https://cdn.yoursite.com/production/datafile-tag-web.json",
+
+    // optional
+    refreshInterval: 3600, // 1 hour
+    timeout: 5, // 5 seconds
+    retry: {
+      // required
+      attempts: 5,
+      interval: 5,
+
+      // optional
+      backOffRate: 2,
+      jitterRange: [0, 5] // 0 and 5 seconds
+    }
+  },
+
+  // SDK setup specific configs
+  launcher: {
+    // url (not needed because this is something we expect app developers to hardcode in their apps)
+
+    // optional
+    refreshInterval: 3600, // 1 hour
+    timeout: 5, // 5 seconds
+    retry: {
+      // required
+      attempts: 5,
+      interval: 5,
+
+      // optional
+      backOffRate: 2,
+      jitterRange: [0, 5] // 0 and 5 seconds
+    }
+  }
+}
+```
+
+### Comparing datafile and launcher config
+
+They are pretty much the same:
+
+| Config key          | Datafile | Launcher   | Description                          |
+| ------------------- | -------- | ---------- | ------------------------------------ |
+| `url`               | Required | Not needed | URL to fetch the JSON file           |
+| `refreshInterval`   | Optional | Optional   | How often to fetch the JSON file     |
+| `timeout`           | Optional | Optional   | Request timeout                      |
+| `retry`             | Optional | Optional   | Retry logic                          |
+| `retry.attempts`    | Optional | Required   | Number of retry attempts             |
+| `retry.interval`    | Optional | Required   | Interval between retries             |
+| `retry.backOffRate` | Optional | Optional   | Exponential backoff rate for retries |
```

Comment:
How the backOff should work?

Reply:
I am undecided if I should also add a new `delay` option here, that could be used along with `backkOffRate`.

imagine an API like:

```js
{
  delay: 500, // ms
  backOffRate: 2, // 2 times
}
```

when retrying, for each retry the waiting period would then be:

- `n` (retry count) * `delay` * `backOffRate`